File: Algorithms/1302_Deepest_Leaves_Sum/Java/DeepestLeavesSumSolution2.java

// Space: O(1)
// Time: O(n)

// DFS approach

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private int res = 0;
    private int deepestLevel = 0;

    public int deepestLeavesSum(TreeNode root) {
        if (root == null) return 0;
        if (root.left == null && root.right == null) return root.val;
        dfs(root, 0);
        return this.res;
    }

    private void dfs(TreeNode node, int level) {
        if (node.left == null && node.right == null) {
            if (level == this.deepestLevel) {
                this.res += node.val;
            } else if (level > this.deepestLevel) {
                this.deepestLevel = level;
                this.res = node.val;
            }
            return;
        }
        if (node.left != null) dfs(node.left, level + 1);
        if (node.right != null) dfs(node.right, level + 1);
    }
}

File: Algorithms/1302_Deepest_Leaves_Sum/Python/Deepest_Leaves_Sum_Solution_2.py

# Space: O(1)
# Time: O(n)

# DFS approach

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def deepestLeavesSum(self, root):
        if root is None: return 0
        if root.left is None and root.right is None: return root.val
        self.res, self.deepest_level = 0, 0

        def dfs(node, level):
            if node.left is None and node.right is None:
                if level == self.deepest_level:
                    self.res += node.val
                elif level > self.deepest_level:
                    self.res = node.val
                    self.deepest_level = level
                return

            if node.left: dfs(node.left, level + 1)
            if node.right: dfs(node.right, level + 1)

        dfs(root, 0)
        return self.res





